server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java

```diff
@@ -171,6 +178,49 @@ public void testEnqueueSpawnsNewTasksUpToMax() throws Exception {
         assertThat(taskRunner.queueSize(), equalTo(0));
     }
 
+    public void testFailsTasksOnRejectionOrShutdown() throws Exception {
+        final var maxThreads = between(1, 5);
+        final var threadFactory = EsExecutors.daemonThreadFactory("test");
+        final var threadContext = new ThreadContext(Settings.EMPTY);
+        final var executor = randomBoolean()
+            ? EsExecutors.newScaling("test", 1, maxThreads, 0, TimeUnit.MILLISECONDS, true, threadFactory, threadContext)
+            : EsExecutors.newFixed("test", maxThreads, between(1, 5), threadFactory, threadContext, false);
+        final var taskRunner = new PrioritizedThrottledTaskRunner<TestTask>("test", between(1, maxThreads * 2), executor);
+        final var totalPermits = between(1, maxThreads * 2);
+        final var permits = new Semaphore(totalPermits);
+        final var taskCompleted = new CountDownLatch(between(1, maxThreads * 2));
+
+        final var spawnThread = new Thread(() -> {
+            try {
+                while (true) {
+                    assertTrue(permits.tryAcquire(10, TimeUnit.SECONDS));
+                    taskRunner.enqueueTask(new TestTask(taskCompleted::countDown, getRandomPriority()) {
+                        @Override
+                        public void onRejection(Exception e) {
+                            // ok
+                        }
+
+                        @Override
+                        public void onAfter() {
+                            permits.release();
+                        }
+                    });
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        });
+        spawnThread.start();
+        assertTrue(taskCompleted.await(10, TimeUnit.SECONDS));
+        executor.shutdown();
+        assertTrue(executor.awaitTermination(30, TimeUnit.SECONDS));
+        spawnThread.interrupt();
+        spawnThread.join();
+        assertThat(taskRunner.runningTasks(), equalTo(0));
+        assertThat(taskRunner.queueSize(), equalTo(0));
```

Comment:
Should we assert at least one rejection happened?

Reply:
Sure, I think we can do that. See 0fc9484ef6493e17649ffa99773237801338e913.